dali/operators/decoder/audio/audio_decoder.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_DECODER_AUDIO_AUDIO_DECODER_H_
+#define DALI_OPERATORS_DECODER_AUDIO_AUDIO_DECODER_H_
+
+#include <memory>
+#include "dali/core/span.h"
+
+namespace dali {
+
+template<typename T>
+struct AudioData {
+  std::shared_ptr<T> data;
+  int length;
+  int sample_rate;  /// [Hz]
+  int channels;
+  bool channels_interleaved;
+};
+
+template<typename T>
+struct AudioDecoder {
+  /**
+   * @return AudioData structure, that has shared ownership of the data.
+   */
+  virtual AudioData<T> Decode(span<const char> encoded) = 0;
+
+  virtual ~AudioDecoder() = default;
+};
+
+template<typename T>
+struct AllocatingDecoder {
+};
+
+template<typename T>
+struct NonallocatingDecoder {
```

Comment:
Why is this called a "decoder" when it doesn't decode anything?

Reply:
That's right - it's not used. I just wanted to provide this possibility in the API, it's not needed now. The most important for me is to have working version. When it's needed, we'll implement